proposals/XXXX-extend-property-wrappers-to-functions-and-closures.md

```diff
@@ -0,0 +1,391 @@
+3. If any of the transformed functions share their signature with another function, the compiler will consider one of the two a redeclaration of the other.
+
+The transformation that will take place is as follows:
+
+1. The parameter name will be prefixed with an underscore.
+2. A synthesized computed property representing  `wrappedValue` will be created and named per the original (non-prefixed) parameter name. The accessors will mirror the `wrappedValue`'s ones.
+3. A synthesized computed property representing  `projectedValue` will be created and named per the original parameter name prefixed with a dollar sign (`$`). The accessors will mirror the `projectedValue`'s ones - except if a mutating setter is defined and the parameter isn't `inout`.
+4. If a parameter's wrapper-type defines an initializer with a first parameter labeled "wrappedValue", then the parameter's label will be prefixed with an underscore and the parameter will be bound to the wrapper type. Also, if that condition applies, an overload will be created with the orginal parameter label and the `wrappedValue`'s type.
+5. If 4 does not apply, then the parameter's label is retained as is and the parameter is bound to the type of `wrappedValue`.
+
+#### Transformation Examples: 
+
+1. Reference Semantics Wrapper
+```swift
+@propertyWrapper
+struct Reference<Value> {
+    init(getter: () -> Value, setter: (Value) -> Void) 
+    
+    var wrappedValue: Value {
+        get 
+        nonmutating set
+    }
+    
+    var projectedValue: Self {
+        self
+    }
+}
+
+func a(@Reference foo: Int) { ... }
+```
+
+    Becomes:
+
+```swift
+func a(foo _foo: Reference<Int>) {
+    var foo: Int {
+        get { 
+            _foo.wrappedValue 
+        }
+        set { 
+            _foo.wrappedValue = newValue
+        }
+    }
+    
+    var $foo: Int {
+        get { 
+            _foo.projectedValue 
+        }
+    }
+    
+    ...
+}
+```
+
+2. Value Semantics Wrapper
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+    init(wrappedValue: Value) 
+    
+    var wrappedValue: Value {
+        get 
+        set
+    }
+}
+
+func b(@Wrapper foo: inout Int) { ... }
+```
+
+    Becomes:
+
+```swift
+func b(_foo: inout Wrapper<Int>) {
+    var foo: Int { ... }
+}
+
+// with the overload:
+
+func b(foo: inout Int) {
+    var _foo: Wrapper<Int> {
+        get {
+            Wrapper(wrappedValue: foo)
+        }
+        set {
+            foo = newValue.wrappedValue
+        }
+    }
+
+    c(_foo: _foo)
+}
+```
+
+3. Reference Semantics Wrapper with Special Initializer
+
+```swift 
+@propertyWrapper
+class WrapperObject<Value> {
+    init(wrappedValue: Value) 
+    
+    var wrappedValue: Value {
+        get 
+        set 
+        // Not actually mutating
+        // because `WrapperObject` is
+        // a class.
+    }
+}
+```
+
+    Becomes:
+
+```swift
+func c(_foo: WrapperObject<Int>) {
+    var foo: Int { 
+        _foo.wrappedValue 
+    }
+}
+
+// with the overload:
```

Comment:
I think it's important here to not show the transformation as the compiler generating overloads of the function, but rather showing how calling the function is transformed to insert the `init(wrappedValue:)` call around the property wrapper (wrapped value) argument, e.g.

```swift
func fn(@WrapperObject value: Int) { ... }

fn(value: 10)
```

Becomes

```swift
func fn(value _value: WrapperObject<Int>) { ... }

fn(value: WrapperObject(wrappedValue: 10))
```

This is how the transformation is implemented, but it also better fits the mental model of initializing a property wrapper from a wrapped value, rather than thinking about it as the compiler choosing an overload.

Reply:
Okey, I've changed that. The only transformation that still has an overload is the `inout` one. Let me know how the version discussed in the proposal seems to you. Otherwise, we can remove any `inout` logic from function parameters and just mention `inout` in a Future Directions section.